compiler/rustc_mir/src/transform/check_consts/qualifs.rs

```diff
@@ -78,18 +155,102 @@ pub struct HasMutInterior;
 impl Qualif for HasMutInterior {
     const ANALYSIS_NAME: &'static str = "flow_has_mut_interior";
 
-    fn in_qualifs(qualifs: &ConstQualifs) -> bool {
-        qualifs.has_mut_interior
+    fn in_qualifs(qualifs: &ConstQualifs) -> Option<()> {
+        qualifs.has_mut_interior.then_some(())
     }
 
-    fn in_any_value_of_ty(cx: &ConstCx<'_, 'tcx>, ty: Ty<'tcx>) -> bool {
-        !ty.is_freeze(cx.tcx.at(DUMMY_SP), cx.param_env)
+    fn in_any_value_of_ty(cx: &ConstCx<'_, 'tcx>, ty: Ty<'tcx>) -> Option<()> {
+        (!ty.is_freeze(cx.tcx.at(DUMMY_SP), cx.param_env)).then_some(())
     }
 
-    fn in_adt_inherently(cx: &ConstCx<'_, 'tcx>, adt: &'tcx AdtDef, _: SubstsRef<'tcx>) -> bool {
+    fn in_adt_inherently(
+        cx: &ConstCx<'_, 'tcx>,
+        adt: &'tcx AdtDef,
+        _: SubstsRef<'tcx>,
+    ) -> Option<()> {
         // Exactly one type, `UnsafeCell`, has the `HasMutInterior` qualif inherently.
         // It arises structurally for all other types.
-        Some(adt.did) == cx.tcx.lang_items().unsafe_cell_type()
+        (Some(adt.did) == cx.tcx.lang_items().unsafe_cell_type()).then_some(())
+    }
+}
+
+/// Constant containing interior mutability (`UnsafeCell<T>`) behind a reference.
+/// This must be ruled out to make sure that evaluating the constant at compile-time
+/// and at *any point* during the run-time would produce the same result. In particular,
+/// promotion of temporaries must not change program behavior; if the promoted could be
+/// written to, that would be a problem.
+pub struct HasMutInteriorBehindRef;
+
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub enum HasMutInteriorBehindRefState {
+    Yes,
+    /// As long as we haven't encountered a reference yet, we use this state
+    /// which is equivalent to the `HasMutInterior` qualif.
```

Comment:
in a follow up PR we can check whether we can get rid of `HasMutInterior` and only run this qualif